### DISCUSSION TITLE
Using debug function with picoprobe, arduino IDE 2.3.2 on windows 10 x64

### DISCUSSION THREAD

**Discussion by nakeze** [earlephilhower/arduino-pico#2297] [General]
Hello,
I am working with Arduino 2.3.2 on a windows 10 x64 machine, and this core  3.1.1.
I am developping using this rp2040 core for a moment now. I would like to use the debugger functionality.
From the doc it seems possible as there is a uploading from picotool, pico-debug, picoprobe.
I have a pi pico and a pi debug probe. 
When I set up the SWD wires between probe and pico board, then connect to usb the two devices so that they are powered, I can see to serial ports in the Arduino IDE (drivers should be good).
However when I select the pico probe port and change the upload method to picoprobe (regular one), I have this message : 
Open On-Chip Debugger 0.11.0-g228ede43d-dirty (2022-12-22-20:09)
Licensed under GNU GPL v2
For bug reports, read
	http://openocd.org/doc/doxygen/bugs.html
Info : only one transport option; autoselect 'swd'
adapter speed: 5000 kHz

Info : Hardware thread awareness created
Info : Hardware thread awareness created
Info : RP2040 Flash Bank Command
Error: Failed to open or find the device
Error: Can't find a picoprobe device! Please check device connections and permissions.
Error: No Valid JTAG Interface Configured.
Failed uploading: uploading error: exit status 0xffffffff

If I select the CMSIS-DAP picoprobe, I have the status 1

I downloaded the pico SDK installer (that installs other softwares, openocd and gdb are in there) and it is not helping.

Is it possible to do what I am trying to achieve ? What am I doing wrong ?
Sincerely
Nicolas

**Comment by earlephilhower** [comment]
Moved to a discussion because this isn't really core related.

I'll assume you bought this from the R.PI guys: https://www.raspberrypi.com/products/debug-probe/

It's a debug-probe but, confusingly, it runs an app which was called PicoProbe for years, but evidently was [renamed "debugprobe" earlier this year for legal reasons](https://github.com/raspberrypi/debugprobe/commit/d0c03d2564fb6aa2cccfe9e08f48efe7e2463ace).  In any case, it's still called "PicoProbe" in the core and OpenOCD.

Thankfully, though, even with the device name changes it is still recognized (just installed [debugprobe 2.0.1](https://github.com/raspberrypi/debugprobe/releases/tag/debugprobe-v2.0.1) on a board that was running much older SW):
````
<under older 1.X SW>
earle@amd:~/Arduino/hardware/pico/rp2040$ ./system/openocd/bin/openocd -f lib/picoprobe_cmsis_dap.tcl 
Open On-Chip Debugger 0.12.0-g4d87f6dca (2023-09-11-23:12)
Licensed under GNU GPL v2
For bug reports, read
	http://openocd.org/doc/doxygen/bugs.html
Info : Hardware thread awareness created
Info : Hardware thread awareness created
Info : Listening on port 6666 for tcl connections
Info : Listening on port 4444 for telnet connections
Info : Using CMSIS-DAPv2 interface with VID:PID=0x2e8a:0x000c, serial=E66038B7134A9339
Info : CMSIS-DAP: SWD supported
Info : CMSIS-DAP: Atomic commands supported
Info : CMSIS-DAP: Test domain timer supported
Info : CMSIS-DAP: FW Version = 2.0.0
Info : CMSIS-DAP: Interface Initialised (SWD)
Info : SWCLK/TCK = 0 SWDIO/TMS = 0 TDI = 0 TDO = 0 nTRST = 0 nRESET = 0
Info : CMSIS-DAP: Interface ready
Info : clock speed 5000 kHz
Error: Failed to connect multidrop rp2040.dap0

<Upgrade to latest @ their repo...note that on the real Pico HW you need to use `debugprobe_on_pico` and not the default `debugprobe` version or you **will** have connection failures>
earle@amd:~/Arduino/hardware/pico/rp2040$ cp ~/Downloads/debugprobe_on_pico.uf2 /media/earle/RPI-RP2/.
earle@amd:~/Arduino/hardware/pico/rp2040$ ./system/openocd/bin/openocd -f lib/picoprobe_cmsis_dap.tcl 
Open On-Chip Debugger 0.12.0-g4d87f6dca (2023-09-11-23:12)
Licensed under GNU GPL v2
For bug reports, read
	http://openocd.org/doc/doxygen/bugs.html
Info : Hardware thread awareness created
Info : Hardware thread awareness created
Info : Listening on port 6666 for tcl connections
Info : Listening on port 4444 for telnet connections
Info : Using CMSIS-DAPv2 interface with VID:PID=0x2e8a:0x000c, serial=E66038B7134A9339
Info : CMSIS-DAP: SWD supported
Info : CMSIS-DAP: Atomic commands supported
Info : CMSIS-DAP: Test domain timer supported
Info : CMSIS-DAP: FW Version = 2.0.0
Info : CMSIS-DAP: Interface Initialised (SWD)
Info : SWCLK/TCK = 0 SWDIO/TMS = 0 TDI = 0 TDO = 0 nTRST = 0 nRESET = 0
Info : CMSIS-DAP: Interface ready
Info : clock speed 5000 kHz
Error: Failed to connect multidrop rp2040.dap0
````

So, the toolchain works with the updated SW stack from R.Pi.

PicoProbe is the name for the Pico-driven debugger as a separate board.  Using anything other than that to talk to the board won't work.  `pico-debug` is [another _completely different_ debug method](https://github.com/majbthrd/pico-debug) that uses no extra HW.  It steals one core of the Pico, some RAM, and the USB port to implement a SWD debugger on the other core (which can only communicate over UART serial Serial1/2 and other peripherals, not USB).  So make sure you're using `PicoProbe` as the upload method.

With that configuration
![image](https://github.com/user-attachments/assets/afe3cd97-5070-435f-b50e-4cd569c2963e)

I get an OK config.
````
Open On-Chip Debugger 0.12.0-g4d87f6dca (2023-09-11-23:12)
Licensed under GNU GPL v2
For bug reports, read
	http://openocd.org/doc/doxygen/bugs.html
Info : Hardware thread awareness created
Info : Hardware thread awareness created
adapter speed: 5000 kHz

Info : Using CMSIS-DAPv2 interface with VID:PID=0x2e8a:0x000c, serial=E66038B7134A9339
Info : CMSIS-DAP: SWD supported
Info : CMSIS-DAP: Atomic commands supported
Info : CMSIS-DAP: Test domain timer supported
Info : CMSIS-DAP: FW Version = 2.0.0
Info : CMSIS-DAP: Interface Initialised (SWD)
Info : SWCLK/TCK = 0 SWDIO/TMS = 0 TDI = 0 TDO = 0 nTRST = 0 nRESET = 0
Info : CMSIS-DAP: Interface ready
Info : clock speed 5000 kHz
Info : SWD DPIDR 0x0bc12477, DLPIDR 0x00000001
Info : SWD DPIDR 0x0bc12477, DLPIDR 0x10000001
Info : [rp2040.core0] Cortex-M0+ r0p1 processor detected
Info : [rp2040.core0] target has 4 breakpoints, 2 watchpoints
Info : [rp2040.core1] Cortex-M0+ r0p1 processor detected
Info : [rp2040.core1] target has 4 breakpoints, 2 watchpoints
Info : starting gdb server for rp2040.core0 on 3333
Info : Listening on port 3333 for gdb connections
[rp2040.core0] halted due to debug-request, current mode: Thread 
xPSR: 0xf1000000 pc: 0x000000ee msp: 0x20041f00
[rp2040.core1] halted due to debug-request, current mode: Thread 
xPSR: 0xf1000000 pc: 0x000000ee msp: 0x20041f00
** Programming Started **
Info : Found flash device 'win w25q16jv' (ID 0x001540ef)
Info : RP2040 B0 Flash Probe: 2097152 bytes @0x10000000, in 32 sectors

Info : Padding image section 1 at 0x10011170 with 144 bytes (bank write end alignment)
Warn : Adding extra erase range, 0x10011200 .. 0x1001ffff
** Programming Finished **
** Verify Started **
** Verified OK **
[rp2040.core0] halted due to debug-request, current mode: Thread 
xPSR: 0xf1000000 pc: 0x000000ee msp: 0x20041f00
[rp2040.core1] halted due to debug-request, current mode: Thread 
xPSR: 0xf1000000 pc: 0x000000ee msp: 0x20041f00
````

Unless they have changed the `DebugProbe` to be incompatible with their original `PicoProbe` software, then I think that's your issue there.

**Reply by earlephilhower** [reply]
Also, is the SDK's versions of all this toolchain (incompatible w/the Arduino-Pico core, but 100% working on its own) working properly?

If you start the SDK's `openocd` with their specified TCL file, do you get the same issue?  If so, then it's something to do with the DebugProbe itself.  You might try reflashing the latest SW from their repo in that case...

**Reply by earlephilhower** [reply]
Also, just noticed, your original post isn't using the core's OpenOCD version which is newer than it and also reported as not stock/git clean:
> ` Open On-Chip Debugger 0.11.0-g228ede43d-dirty (2022-12-22-20:09)`

vs.
> `Open On-Chip Debugger 0.12.0-g4d87f6dca (2023-09-11-23:12)`
